From fb3d0e6706d9913ae4c64065dc66d4dd44eaeb14 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 19 Mar 2025 11:59:27 +0100
Subject: [PATCH 1643/1695] plat-stm32mp2: disable parent IRQ only when all PWR
 client disable it

The PWR has till 6 interrupt clients and propagate them to a
single interrupt to the parent interrupt controller.

Keep bitmasks of enabled and masked client interrupts.
Enable the parent interrupt as soon as a client asks enable.
Disable the parent interrupt when no more clients enable it.
Mask the parent interrupt as soon as one client asks mask.
Unmask the parent interrupt when no more clients mask it.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 49477934be6f95d1c9a56fa259a15368bd4ba732 plat-stm32mp2: disable parent IRQ only when all PWR client disable it]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I72d5b673b305721229cd36532148ad349fa695aa
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/445232
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 .../plat-stm32mp2/drivers/stm32mp25_pwr_irq.c | 21 ++++++++++++-------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
index 9837bca78..a95e8c91c 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
@@ -53,6 +53,8 @@ struct pwr_wkup_data {
 	struct itr_handler *parent_hdl;
 	struct stm32_exti_pdata *exti;
 	unsigned int spinlock;
+	uint8_t itr_enable_bitmask;
+	uint8_t itr_mask_bitmask;
 };
 
 static struct pwr_wkup_data *pwr_wkup_d;
@@ -197,10 +199,10 @@ static void stm32mp25_pwr_op_enable(struct itr_chip *chip __unused, size_t it)
 	uint32_t exceptions = 0;
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	priv->itr_enable_bitmask |= BIT(it);
 	stm32mp25_pwr_itr_enable_nolock(it);
-	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-
 	interrupt_enable(priv->parent_hdl->chip, priv->parent_hdl->it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
 }
 
 /* Disable an interrupt */
@@ -210,10 +212,12 @@ static void stm32mp25_pwr_op_disable(struct itr_chip *chip __unused, size_t it)
 	uint32_t exceptions = 0;
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	priv->itr_enable_bitmask &= ~BIT(it);
 	stm32mp25_pwr_itr_disable_nolock(it);
+	if (!priv->itr_enable_bitmask)
+		interrupt_disable(priv->parent_hdl->chip,
+				  priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-
-	interrupt_disable(priv->parent_hdl->chip, priv->parent_hdl->it);
 }
 
 /* Mask an interrupt, may be called from an interrupt context */
@@ -223,10 +227,10 @@ static void stm32mp25_pwr_op_mask(struct itr_chip *chip __unused, size_t it)
 	uint32_t exceptions = 0;
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	priv->itr_mask_bitmask |= BIT(it);
 	stm32mp25_pwr_itr_disable_nolock(it);
-	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-
 	interrupt_mask(priv->parent_hdl->chip, priv->parent_hdl->it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
 }
 
 /* Unmask an interrupt, may be called from an interrupt context */
@@ -236,10 +240,11 @@ static void stm32mp25_pwr_op_unmask(struct itr_chip *chip __unused, size_t it)
 	uint32_t exceptions = 0;
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	priv->itr_mask_bitmask &= ~BIT(it);
 	stm32mp25_pwr_itr_enable_nolock(it);
+	if (!priv->itr_mask_bitmask)
+		interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-
-	interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
 }
 
 /* Raise per-cpu interrupt (optional) */
-- 
2.47.2

